src/main/scala/is/hail/io/RowStore.scala

```diff
@@ -521,7 +523,95 @@ object RichRDDRegionValue {
 }
 
 class RichRDDRegionValue(val rdd: RDD[RegionValue]) extends AnyVal {
-  def writeRows(path: String, t: TStruct): Array[Long] = {
+  def writeRows(path: String, t: TStruct): (Array[String], Array[Long]) = {
     rdd.writePartitions(path, RichRDDRegionValue.writeRowsPartition(t))
   }
+
+  def writeRowsSplit(path: String, t: MatrixType, partitioner: OrderedRVDPartitioner): Array[Long] = {
+    val sc = rdd.sparkContext
+    val hConf = sc.hadoopConfiguration
+
+    hConf.mkDir(path + "/rows/rows/parts")
+    hConf.mkDir(path + "/entries/rows/parts")
+
+    val sHConfBc = sc.broadcast(new SerializableHadoopConfiguration(hConf))
+
+    val nPartitions = rdd.getNumPartitions
+    val d = digitsNeeded(nPartitions)
+
+    val partFiles = Array.tabulate[String](nPartitions) { i => partFile(d, i) }
+
+    val fullRowType = t.rvRowType
+    val rowsRVType = t.rowType
+    val localEntriesIndex: Int = t.entriesIdx
+
+    val entriesRVType = TStruct(
+      MatrixType.entriesIdentifier -> TArray(t.entryType))
+
+    val partitionCounts = rdd.mapPartitionsWithIndex { case (i, it) =>
+      val hConf = sHConfBc.value.value
+
+      val f = partFile(d, i)
+
+      val rowsPartPath = path + "/rows/rows/parts/" + f
+      val rowsOS = hConf.unsafeWriter(rowsPartPath)
+
+      val entriesPartPath = path + "/entries/rows/parts/" + f
+      val entriesOS = sHConfBc.value.value.unsafeWriter(entriesPartPath)
+
+      val rowsEN = new Encoder(new LZ4OutputBuffer(rowsOS))
+      val entriesEN = new Encoder(new LZ4OutputBuffer(entriesOS))
+      var rowCount = 0L
+
+      val rvb = new RegionValueBuilder()
+      val fullRow = new UnsafeRow(fullRowType)
+
+      it.foreach { rv =>
+        fullRow.set(rv)
+        val row = fullRow.deleteField(localEntriesIndex)
+
+        val region = rv.region
+        rvb.set(region)
+        rvb.start(rowsRVType)
+        rvb.addAnnotation(rowsRVType, row)
+
+        rowsEN.writeByte(1)
+        rowsEN.writeRegionValue(rowsRVType, region, rvb.end())
+
+        rvb.start(entriesRVType)
+        rvb.startStruct()
+        rvb.addField(fullRowType, rv, localEntriesIndex)
+        rvb.endStruct()
+
+        entriesEN.writeByte(1)
+        entriesEN.writeRegionValue(entriesRVType, region, rvb.end())
+
+        rowCount += 1
+      }
+
+      rowsEN.writeByte(0) // end
+      rowsEN.flush()
+      rowsOS.close()
+
+      entriesEN.writeByte(0)
+      entriesEN.flush()
+      entriesOS.close()
```

Comment:
these really ought to be in `try ... finally` blocks to ensure we're always closing the output streams. The Oracle standard library tends to burden the containing class with resource management. Under that theory, Encoder and friends should be Closable and have their close method just call their childrens' close.